### PR TITLE
Compare symbol names when filtering duplicates in cse

### LINKED ISSUE #25043: Lambdify with cse wrong result
Using `lambdify` with `cse=True` sometimes returns the wrong result.
Minimal example:
```python
import sympy as sp
c = sp.symbols("c")
x = sp.symbols("x0", real=True)
poly = c*x**2 + c*(x**4 - x**2)

f = sp.lambdify((c, x), poly, cse=True)  #  <- `cse=True` makes the difference
g = sp.lambdify((c, x), poly, cse=False)

#  wrong result, AssertionError: 256 /= 16
assert f(1, 2) == g(1, 2), f"{f(1, 2)} /= {g(1, 2)}"
```

**Cause**
The issue seems to be related to the filtering of symbols to be excluded in the `tree_cse` function around here [https://github.com/sympy/sympy/blob/master/sympy/simplify/cse_main.py#L655](https://github.com/sympy/sympy/blob/53a36aa71956c988e2b6ce33db0d8cbcbdfacfbb/sympy/simplify/cse_main.py#L655).

From what I can tell, the cse method aims to avoid name collisions between symbols already in the expression and the symbols representing common subexpressions.
This is done by adding all symbols found in the expression to an `excluded_symbols` set and filtering the `symbols` iterator to exclude them.
However, symbols are equal iff their names and assumptions match, so this approach does not avoid symbol *name collisions*:

```python
from sympy import Symbol
Symbol("x") in {Symbol("x", real=True)}  # returns False
```

**Suggested solution**
In the `tree_cse` function, instead of filtering `symbols` from all symbols occuring in the expression, 
filter from all symbols sharing `name` with some symbol in the expression.

If this can be considered a sufficient fix for the issue I'd be happy to provide a PR.

### PR DESCRIPTION
**Pull request by jokasimr**
The cse method aims to avoid name collisions between symbols already in the expression and the symbols representing common subexpressions.
This is done by adding all symbols found in the expression to an `excluded_symbols` set and filtering the symbols iterator to exclude them.
However, symbols are equal only if their names and assumptions match, so this approach does not avoid symbol name collisions.

#### References to other Issues or PRs
Fixes #25043 

#### Brief description of what is fixed or changed
`excluded_symbols` stores names of symbols instead of symbols.
The symbols iterator is filtered to exlude all symbols sharing a name with a symbols already in the expression.

#### Release Notes

* simplify
    * Fixed a bug where cse adds temporary variables sharing names with symbols already in the expression.

### CODE REVIEW

**Comment by smichr** [comment]
This has no impact on behavior but would lead to a less ambiguous represenation of the replacements for the human reader.

File: sympy/simplify/cse_main.py [hunk 1]
```diff
@@ -652,7 +652,7 @@ def _find_repeated(expr):
     ## Rebuild tree
 
     # Remove symbols from the generator that conflict with names in the expressions.
-    symbols = (symbol for symbol in symbols if symbol not in excluded_symbols)
+    symbols = (symbol for symbol in symbols if str(symbol) not in excluded_symbols)
```
Comment (smichr) [inline sympy/simplify/cse_main.py:655]:
```suggestion
    symbols = (s for s in symbols if s.name not in excluded_symbols)
```

File: sympy/simplify/cse_main.py [hunk 2]
```diff
@@ -621,7 +621,7 @@ def _find_repeated(expr):
                 expr.is_Order or
                 isinstance(expr, (MatrixSymbol, MatrixElement))):
             if expr.is_Symbol:
-                excluded_symbols.add(expr)
+                excluded_symbols.add(str(expr))
```
Comment (smichr) [inline sympy/simplify/cse_main.py:624]:
```suggestion
                excluded_symbols.add(expr.name)
```

File: sympy/simplify/tests/test_cse.py [hunk 3]
```diff
@@ -246,6 +247,13 @@ def test_issue_6263():
     assert cse(e, optimizations='basic') == ([], [True])
 
 
+def test_issue_25043():
+    c = symbols("c")
+    x = symbols("x0", real=True)
+    cse_expr = cse(c*x**2 + c*(x**4 - x**2))[-1][-1]
+    assert Counter((s.name for s in cse_expr.free_symbols))["x0"] == 1
```
Comment (smichr) [inline sympy/simplify/tests/test_cse.py:254]:
```suggestion
    free = cse_expr.free_symbols
    assert len(free) == len(set(i.name for i in free))
```

Comment (smichr) [reply sympy/simplify/tests/test_cse.py:254]: (And then Counter import could be removed.)

**Comment by smichr** [comment]
Less visually confusing cse output has been committed, thanks!

**Comment by jokasimr** [comment]
Just to be clear, it's not visually confusing but also a cause of bugs. Example:
```python
import sympy as sp
c = sp.symbols("c")
x = sp.symbols("x0", real=True)
poly = c*x**2 + c*(x**4 - x**2)

f = sp.lambdify((c, x), poly, cse=True)  #  <- `cse=True` makes the difference
g = sp.lambdify((c, x), poly, cse=False)

#  wrong result, 256 /= 16
print(f(1, 2), g(1, 2))
```

**Pull request merged**